Sources/SwiftProtobuf/Google_Protobuf_FieldMask+Extensions.swift

```diff
@@ -184,3 +179,201 @@ extension Google_Protobuf_FieldMask: _CustomJSONCodable {
+  ///      is also in the FieldMask.
+  ///   2. Sort all paths in alphabetical order.
+  public var canonical: Google_Protobuf_FieldMask {
+    var mask = Google_Protobuf_FieldMask()
+    let sortedPaths = self.paths.sorted()
+    var set = Set<String>()
+    for path in sortedPaths {
+      if !contains(path, in: set) {
+        set.insert(path)
+        mask.paths.append(path)
+      }
+    }
+    return mask
+  }
+
+  /// Creates an union of two FieldMasks.
+  ///
+  /// - Parameter mask: FieldMask to union with.
+  /// - Returns: FieldMask with union of two path sets.
+  public func union(
+    _ mask: Google_Protobuf_FieldMask
+  ) -> Google_Protobuf_FieldMask {
+    var buffer: Set<String> = .init()
+    var _paths: [String] = []
+    let allPaths = paths + mask.paths
+    for path in allPaths where !buffer.contains(path) {
+      buffer.insert(path)
+      _paths.append(path)
+    }
+    return .with { mask in
+      mask.paths = _paths
+    }
+  }
+
+  /// Creates an intersection of two FieldMasks.
+  ///
+  /// - Parameter mask: FieldMask to intersect with.
+  /// - Returns: FieldMask with intersection of two path sets.
+  public func intersect(
+    _ mask: Google_Protobuf_FieldMask
+  ) -> Google_Protobuf_FieldMask {
+    let set = mask.pathsSet
+    var _paths: [String] = []
+    for path in paths where set.contains(path) {
+      _paths.append(path)
+    }
+    return .with { mask in
+      mask.paths = _paths
+    }
+  }
+
+  /// Creates a FieldMasks with paths of the original FieldMask
+  /// that does not included in mask.
+  ///
+  /// - Parameter mask: FieldMask with paths should be substracted.
+  /// - Returns: FieldMask with all paths does not included in mask.
+  public func subtract(
+    _ mask: Google_Protobuf_FieldMask
+  ) -> Google_Protobuf_FieldMask {
+    let set = mask.pathsSet
+    var _paths: [String] = []
+    for path in paths where !set.contains(path) {
+      _paths.append(path)
+    }
+    return .with { mask in
+      mask.paths = _paths
+    }
+  }
+
+  /// Returns true if path is covered by the given FieldMask. Note that path
+  /// "foo.bar" covers all paths like "foo.bar.baz", "foo.bar.quz.x", etc.
+  /// Also note that parent paths are not covered by explicit child path, i.e.
+  /// "foo.bar" does NOT cover "foo", even if "bar" is the only child.
+  ///
+  /// - Parameter path: Path to be checked.
+  /// - Returns: Boolean determines is path covered.
+  public func contains(_ path: String) -> Bool {
+    contains(path, in: pathsSet)
+  }
+
+  // Set containing paths of FieldMask
+  private var pathsSet: Set<String> {
+    .init(paths)
+  }
+
+  private func levels(path: String) -> [String] {
+    let comps = path.components(separatedBy: ".")
+    return (0..<comps.count).map {
+      comps[0...$0].joined(separator: ".")
+    }
+  }
+
+  private func contains(
+    _ path: String,
+    in set: Set<String>
+  ) -> Bool {
+    for level in levels(path: path) {
```

Comment:
Also a little worried about memory usage/performance here, might make sense to use string ops to keep things in check.

Reply:
With the algorithm you've mentioned above, we won't need it anymore